### ISSUE TITLE
Support streams or binary data

### ISSUE THREAD

**Issue by rumbu13** [homieiot/convention#287]
Currently Homie Convention allows only UTF8 encoded strings, but in order to keep the pace with the current devices, I thing there is a need for a binary data type.

We have cameras, speakers, players. Devices processing images, sound and voice are everywhere, therefore the most efficient way is to use a binary stream. Of course any binary stream can be base64-encoded and fallback to the string representation but this is far from efficient.

**Comment by Tieske** [comment]
I have no experience with it whatsoever, but how would a binary stream be passed around in MQTT?

**Comment by rumbu13** [comment]
In its core, mqtt works with binary data only, the fact that you choose to encode/decode that data in a string using utf8 is just a choice, mqqt protocol is not aware of that. Therefore the default data format you send/receive to/from a mqtt server are just simple bytes that you obtain through encoding/decoding a string.

That's why we can specify in the convention a new data type called "binary" or "bytes" where you can publish/subscribe raw bytes. It means basically that you don't need to utf8 decode the data found on a subscribed topic and take it as it is. It can be a picture, a sound, a video frame, a zip stream or even raw data captured from a device. The only limit imposed by mqtt protocol is the size of the packet - 256 MB.

What we can do in the convention is to make mandatory to publish also the mime type like image/jpeg, audio/vorbis, font/ttf. More than that, mqtt 5.0 allows to specify the mime type/format in the packet header.

**Comment by Tieske** [comment]
The binary data I get. But you mentioned supporting streams. How would that work?

**Comment by rumbu13** [comment]
Streams are a particular case of binary data, publishing a video frame at variable time intervals makes it a stream. What can be added for this particular case is for example a mandatory empty payload to signal the end of stream for subscribers.

For my particular use case (camera and voice), a binary data type in the convention should be enough.

**Comment by Tieske** [comment]
A binary type does make sense to me as well. You could propose a PR against the [homie5](https://github.com/homieiot/convention/tree/homie5) branch.

**Comment by Thalhammer** [comment]
While I agree with a binary type being a good idea, I don't think the use cases presented here are a good example. MQTT is binary safe, but its not a streaming protocol. Messages can get dropped (at QOS 0) or arrive multiple times in different orders (QOS 1). The maximum message size for a MQTT payload is ~250MB, but the broker is free to impose a lower limit. Most brokers do (AWS: 128Kb, Google Pub/Sub: 256Kb, IBM: 4Mb) and even if you run a broker that allows large payloads you start to hit another problem. MQTT doesn't support splitting the payload, meaning for the duration of the media upload there is no way for you to react to anything else. To make matters worse, you can't even cancel the upload without reconnecting.

TLDR: MQTT is not the right protocol for media streaming. It might work for small deployments and in your special case, but its definitely not going to work at scale. Use a protocol thats designed for that purpose instead (e.g. WebRTC) and exchange the data related to it via mqtt (e.g. peer information / endpoint infos).

**Comment by jacoscaz** [comment]
> TLDR: MQTT is not the right protocol for media streaming. It might work for small deployments and in your special case, but it's definitely not going to work at scale. Use a protocol thats designed for that purpose instead (e.g. WebRTC) and exchange the data related to it via mqtt (e.g. peer information / endpoint infos).

This resonates strongly with my experience. It would be better to extend homie to model external data sources.

**Comment by dcj** [comment]
I concur that attempting to do streaming over MQTT is (sorry to be harsh) a terrible idea.

But, supporting topic values as binary data would/could be very valuable/important.
This would be a signficant change for Homie, and would require careful thought, design, and consensus,

Some quick brainstorm ideas (off the cuff, not vetted):
* Extend/enhance the `$description` schema for a property to enable specifying: value is a binary object of type foo, where foo might be byte8 (or longer?), int32, int64, float32, float64, etc.
* Crazier still: value is a protobuf (or cbor, or even JSON object!) and the schema is bar

I fully expect howls of derision and ridicule in reponse/reply but I have seen real-word needs that something like this would meet...

**Comment by Thalhammer** [comment]
> I fully expect howls of derision and ridicule in reponse/reply

Well then you are probably in for a surprise ;)
Personally I would have loved to move to protobuf for all/most of the stuff in homie. Its a simple, elegant and easy to use encoding and is pretty easy to implement from scratch in less than 500 lines of C/C++ if you restrict yourself to a sensible subset. At this point the problem really is legacy. Moving to protobuf by design is a breaking change and requires major changes to all devices and controllers. Homie 5.0 would have been an option to do so, however I have been overvoted.

Maybe we will see that shift for Homie 6.0 ;)

**Comment by dcj** [comment]
I'm not suggesting that all topic values become protobufs, only that there is a way to specify a datatype other than "string" for a topic value, and if we did that, then datatypes like protobuf might be supportable.

**Comment by Tieske** [comment]
Reading back a little, I think the consensus is that a binary type would be a valuable addition.

All we need is someone that has the use case to make an effort to write a PR.

**Comment by dcj** [comment]
@Tieske Thank you for "confirming" that a binary type proposal would be considered (and not rejected outright on principle)
I will give this some thought & contemplation, but it could easily be a couple of months, so don't hold your breath waiting!

> All we need is someone that has the use case to make an effort to write a PR.

Absolutely!
The Clojure community defines and "strongly urges" an [Etiquette](https://clojure.org/community/etiquette) which I strongly support, one key tenent of which is:

> If you think something ought to be made, then make it. 
> Otherwise, respect others peoples' right to choose what they do with their time.

**Comment by schaze** [comment]
Chiming in because I think it is an interesting topic  (and I still feel a bit guilty that we overruled @Thalhammer  with the JSON vs protobuf discussion - although I still believe it was the right choice - sorry ;-)  ):

Is the binary type suggestion comparable from a use case perspective with the `json` type field? Meaning we have a binary represention of a complex object or data?

**Comment by Tieske** [comment]
the only thing I can come up with is arbitrary binary data. Could be an image, eg video still. Or just some custom encoding. Would be nice if it could carry some type info, eg mime-type or similar (both in the definition, as well as in the payload).

**Comment by jacoscaz** [comment]
TL;DR: +1 to binary payloads if accompanied by mime-type. 

One of the beautiful aspects of Homie is that it's easy. Easy to implement, easy to debug, easy to grasp, easy to develop for. Binary payloads everywhere would significantly affect its ease of adoption. 

As for binary types, I do think it's a good idea but only if accompanied by the mime type as proposed by @Tieske . Anything freeform fosters incompatibility between implementations, which is something we should (IMHO) do our very best to avoid.

**Comment by dcj** [comment]
Again, IMO text/ASCII is just fine for a lot of use cases, and it is super easy.
But, it might be nice if there was an option for binary data, for times when that would be an improvement.
For an IoT device that doesn't change state a lot, just use strings.
For a electrical panel that is publishing ~100 metering measurements per second, I suspect that avoiding the string serdes might be advantageous.
But, I suggest that this hypothesis be tested/proven, with real performance data, and not simply assume that binary is better

**Comment by jacoscaz** [comment]
@dcj I think that would result in an inversion of priority.

On one side I agree that binary (de)serialization can be made more performant than string (de)serialization, though one should never underestimate the level of optimization that some common form of serialization have reached. 

But, IMHO, even assuming binary serialization to be the fastest in any and all cases (which it is not), that level of performance is not something that Homie should prioritise for. At least not at the expense of interoperability and ease of adoption.

Homie is fundamentally not engineered for the exchange of high-frequency data. And that's fine. In fact, it is brilliant. It is engineered for simplicity, for ease of adoption, for inspectability. It lacks a lot of things that are required for the efficient exchange of high-frequency data. For example, batch transfers of timeseries data, which (as far as my experience goes) tend to be orders-of-magnitude more effective than _any_ serialization format applied to individual values. 

In most cases, even plain old CSV scales dramatically better than single-value binary serializations when it comes to shuffling high-frequency data across boundaries.

IMHO, introducing alternative serializations for supported formats would not meaningfully help when it comes to dealing with high-frequency data in any production capacity.

However, Introducing multiple ways to serialise the same datatype (numbers) would absolutely introduce ambiguity, which would directly affects ease of adoption. One device might choose to publish power in binary format, another might decide to publish it as a plain-text number. How would we even draft profiles in this case? What should a client expect?

Here's my opinion after some tuning based on this exchange: yes to binary payloads, only if accompanied by mime type, but not as an alternative to currently-supported datatypes.

**Comment by Tieske** [comment]
> For a electrical panel that is publishing ~100 metering measurements per second, I suspect that avoiding the string serdes might be advantageous.


I think that this will not hold. Simply creating the mqtt message with all its fields and settings already involves a lot of data mangling. I doubt adding a single conversion from binary to ascii will have a real impact. My 2cts.

**Comment by Thalhammer** [comment]
> I still feel a bit guilty that we overruled @Thalhammer with the JSON vs protobuf discussion

No reason for guilt, thats how democracy works ;)

> For a electrical panel that is publishing ~100 metering measurements per second, I suspect that avoiding the string serdes might be advantageous.

Advantageous yes, but in practice probably irrelevant compared to the overhead of running a whole tcp/ip stack with mqtt. The bigger benefit in such a case is likely that you can pack all of those 100 values into a single topic and publish them all at once. This gives you a consistent state and heavily reduces the network traffic. That being said, this is not really something homie is designed for. Everything that exceeds a couple messages per second is not really suited for mqtt in general*.

If you need data like that in real time you are better of with something like modbus (its the standard for industrial control for a reason).

If you only need it for something like history you might be better of packing them into a single string and publishing that. String serdes (on the amount we are talking here) is basically noise compared to the rest of the overhead involved.

* My EV Charger actually does this. I has like 40 state topics that all get updated in real time and post strings, however it only updates them whenever they change, which breaks those 40 updates down to ~5-10 per second.
